src/Tasks/ManifestUtil/LauncherBuilder.cs

```diff
@@ -0,0 +1,131 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.ComponentModel;
+using System.IO;
+using System.Runtime.InteropServices;
+using System.Xml.Serialization;
+using Microsoft.Build.Shared.FileSystem;
+using Microsoft.Build.Tasks.Deployment.Bootstrapper;
+
+namespace Microsoft.Build.Tasks.Deployment.ManifestUtilities
+{
+    /// <summary>
+    /// Adds Launcher and updates its resource
+    /// </summary>
+    [ComVisible(true)]
+    public class LauncherBuilder
+    {
+        private const string LAUNCHER_RESOURCENAME = "FILENAME";
+        private const int LAUNCHER_RESOURCE_TABLE = 50;
+
+        private BuildResults _results;
+
+        public LauncherBuilder(string launcherPath)
+        {
+            LauncherPath = launcherPath;
+        }
+
+        /// <summary>
+        /// Specifies the location of the required Launcher files.
+        /// </summary>
+        /// <value>Path to Launcher files.</value>
+        public string LauncherPath { get; set; }
+
+        public BuildResults Build(string filename, string outputPath)
+        {
+            string launcherFilename = Path.GetFileName(LauncherPath);
+
+            _results = new BuildResults();
+
+            try
+            {
+                if (filename == null)
+                {
+                    _results.AddMessage(BuildMessage.CreateMessage(BuildMessageSeverity.Error, "GenerateLauncher.InvalidInput"));
+                    return _results;
+                }
+
+                if (String.IsNullOrEmpty(outputPath))
+                {
+                    _results.AddMessage(BuildMessage.CreateMessage(BuildMessageSeverity.Error, "GenerateLauncher.NoOutputPath"));
+                    return _results;
+                }
+
+                // Copy setup.bin to the output directory
+                string strOutputExe = System.IO.Path.Combine(outputPath, launcherFilename);
+                if (!CopyLauncherToOutputDirectory(strOutputExe))
+                {
+                    // Appropriate messages should have been stuffed into the results already
+                    return _results;
+                }
+
+                var resourceUpdater = new ResourceUpdater();
+                resourceUpdater.AddStringResource(LAUNCHER_RESOURCE_TABLE, LAUNCHER_RESOURCENAME, filename);
+                if (!resourceUpdater.UpdateResources(strOutputExe, _results))
+                {
+                    return _results;
+                }
+
+                _results.SetKeyFile(launcherFilename);
+                _results.BuildSucceeded();
+            }
+            catch (Exception ex)
+            {
+                _results.AddMessage(BuildMessage.CreateMessage(BuildMessageSeverity.Error, "GenerateLauncher.General", ex.Message));
```

Comment:
Are you sure you want to strip the stack trace?
```suggestion
                _results.AddMessage(BuildMessage.CreateMessage(BuildMessageSeverity.Error, "GenerateLauncher.General", ex));
```

Reply:
Speaking for Nikola since he is OOF, I think this is to be consistent with how we display the build error messages in other places.